### PR TITLE
Support for cancellation and explicit disregard of callback context

### PR DESCRIPTION
**Pull request by bschmidtbauer**
Added an overload to each async operation on `IConfluentClient` that now
takes a `CancellationToken` so that the caller can cancel any in progress
requests. *Note that the current behavior in the `SendRequest` method
converts the `TaskCancelledException` into a
`ConfluentApiUnavailableException` where it states (via the message) that
it was due to timeout. I changed this to allow an explicit user
initiated cancel to propagate the `TaskCanceledException` and the timeout
will continue to convert to the confluent exception.

Added an explicit `ConfigureAwait(false)` call on all awaited calls - this
will help to ensure that the callback doesn't cause a deadlock if the
caller has requirements on callback context (e.g. GUI apps, ASP.NET)
https://msdn.microsoft.com/en-us/magazine/jj991977.aspx

### CODE REVIEW

**Comment by josephjeganathan** [comment]
Hi @bschmidtbauer,

This is a very valid PR, I'll take a look at it over the weekend. 

Thanks

**Comment by bschmidtbauer** [comment]
Thanks

File: src/Confluent.RestClient/ConfluentClient.cs [hunk 1]
```diff
@@ -108,13 +151,22 @@ public ConfluentClient(IConfluentClientSettings clientSettings)
             string topic,
             int? maxBytes = null)
         {
+            return await ConsumeAsBinaryAsync(consumerInstance, topic, CancellationToken.None, maxBytes).ConfigureAwait(false);
+        }
+
+        public async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
+            ConsumerInstance consumerInstance,
+            string topic,
+            CancellationToken cancellationToken,
```
Comment (josephjeganathan) [inline src/Confluent.RestClient/ConfluentClient.cs]: Let's have `CancellationToken cancellationToken` as a last parameter.

File: src/Confluent.RestClient/ConfluentClient.cs [hunk 2]
```diff
@@ -124,40 +176,61 @@ public ConfluentClient(IConfluentClientSettings clientSettings)
             where TKey : class
             where TValue : class
         {
+            return await ConsumeAsAvroAsync<TKey, TValue>(consumerInstance, topic, CancellationToken.None, maxBytes).ConfigureAwait(false);
+        }
+
+        public async Task<ConfluentResponse<List<AvroMessage<TKey, TValue>>>> ConsumeAsAvroAsync<TKey, TValue>(
+            ConsumerInstance consumerInstance,
+            string topic,
+            CancellationToken cancellationToken,
```
Comment (josephjeganathan) [inline src/Confluent.RestClient/ConfluentClient.cs]: Same here

File: src/Confluent.RestClient/IConfluentClient.cs [hunk 3]
```diff
@@ -89,6 +164,21 @@ public interface IConfluentClient : IDisposable
...
+        /// <param name="consumerInstance">Consumer instance</param>
+        /// <param name="topic">The topic to consume messages from</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation</param>
+        /// <param name="maxBytes">The maximum number of bytes of un-encoded keys and values that should be included in the response</param>
+        /// <returns>List of Binary messages</returns>
+        Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
+            ConsumerInstance consumerInstance,
+            string topic,
+            CancellationToken cancellationToken,
+            int? maxBytes = null);
```
Comment (josephjeganathan) [inline src/Confluent.RestClient/IConfluentClient.cs]: Let's have CancellationToken cancellationToken as a last parameter.

This implies that we'll have 4 methods to replace this.

``` C#
        Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
            ConsumerInstance consumerInstance,
            string topic);

        Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
            ConsumerInstance consumerInstance,
            string topic,
            CancellationToken cancellationToken);

        Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
            ConsumerInstance consumerInstance,
            string topic,
            int maxBytes);

        Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
            ConsumerInstance consumerInstance,
            string topic,
            int maxBytes,
            CancellationToken cancellationToken);
```

In the actual implementation (i.e. `ConfluentClient`) we can have all the methods calling to one private helper which is already there.

``` C#
        public async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
            ConsumerInstance consumerInstance,
            string topic,
            CancellationToken cancellationToken,
            int? maxBytes = null)
        {
            string requestUri = BuildConsumeRequestUri(topic, maxBytes);

            // ...
        }
```

Comment (josephjeganathan) [inline src/Confluent.RestClient/IConfluentClient.cs]: Do the same for all the other methods which has `int? maxBytes = null` as an optional and last parameter.

**Comment by josephjeganathan** [comment]
Everything else looks good to me.

**Comment by bschmidtbauer** [comment]
Yep, I will do that.  The reason I didn't have it as the last parameter was
the optional maxBytes parameter so it wasn't possible without either
breaking compatibility or adding overloads that did not allow for
optionally passing maxBytes.

With the overloads you suggested it will break compatibility - just want to
make sure you are ok with that.  Basically anyone who relied on the default
value will break when they recompile.  I'm totally in favor of removing the
optional nature of that parameter but just wanted to make sure.

**Comment by josephjeganathan** [comment]
Yes you're totally right on the compatibility issue here (I should have done with overloads at the first place). 

It's a good practice to have the cancellation token as the last parameter (since it's optional).

I would rather break the compatibility on this occasion. Anyone already have the previous version will continue to work fine and people who update _may_ have this compatibility issue (minimal chance as it's optional already).

I'll add a release note and bump up the minor version once this is changed.

Thanks

**Comment by josephjeganathan** [comment]
By the way, according to mscorlib `CancellationToken.None` is equal to `default(CancellationToken)`

http://referencesource.microsoft.com/#mscorlib/system/threading/CancellationToken.cs,64

Refer this too: https://blogs.msdn.microsoft.com/andrewarnottms/2014/03/19/recommended-patterns-for-cancellationtoken/

We can just keep the existing interface as it is and just add `cancellationToken = default(CancellationToken)` at the end of all the methods and implement it accordingly.

Thus, `int? maxBytes` can stay as it is.

**Comment by bschmidtbauer** [comment]
Yeah - I looked at that.  I wanted to avoid adding more optional args
though.  I'm open to either one but would vote for breaking compatibility
in this case.

I'll leave it up to you - either way is trivial to implement and would also
be easy to resolve on the client side had they relied on the default.

**Comment by josephjeganathan** [comment]
Even though it's documented in mscorlib `CancellationToken.None` = `default(CancellationToken)` it's their very internal implementation and it's not really advisable to use it.

Let's do as you suggested and as commented at the beginning.
- Let's have overloads (no defaults)
- Keep the cancellation token as the very last parameter
- Change `int? maxBytes` into overloads too - https://github.com/josephjeganathan/Confluent.RestClient/pull/6#discussion_r47566980

Many Thanks

Closer to midnight in UK, I hope to merge it tomorrow :)

**Comment by bschmidtbauer** [comment]
I have updated my PR.  Note that I left the maxBytes as a `nullable<int>` but forced it to be a required parameter instead of optional.

The reason for leaving it nullable was it felt more logical the way you had it with being able to check for null to determine if you should supply a parameter to the REST API.  To change it to an `int` would require a conversion somewhere along the way or testing for `<= 0` in order to determine if you should send the parameter to the HTTP endpoint.  I tried it both ways and the `nullable<int>` felt more natural than having to check the int value to determine what to do.  Leaving it as a nullable also allowed the existing test methods to continue to function e.g. `GetMaxBytesOrNull` continued to work as expected vs. having to change all uses of it to determine which overload they should call (with or without the maxBytes as an int).

Let me know your thoughts.

Thanks for doing this.

**Comment by josephjeganathan** [comment]
I would prefer to have the following method in this interface instead.

``` C#
Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(ConsumerInstance consumerInstance, string topic);
Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(ConsumerInstance consumerInstance, string topic, CancellationToken cancellationToken);
Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(ConsumerInstance consumerInstance, string topic, int maxBytes);
Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(ConsumerInstance consumerInstance, string topic, int maxBytes, CancellationToken cancellationToken);
```

For example, you can use the bellow to use a common helper

``` C#
public async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
    ConsumerInstance consumerInstance,
    string topic)
{
    return await ConsumeAsBinaryAsync(consumerInstance, topic, null, CancellationToken.None).ConfigureAwait(false);
}

public async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
    ConsumerInstance consumerInstance,
    string topic,
    CancellationToken cancellationToken)
{
    return await ConsumeAsBinaryAsync(consumerInstance, topic, null, cancellationToken).ConfigureAwait(false);
}

public async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
    ConsumerInstance consumerInstance,
    string topic,
    int maxBytes)
{
    return await ConsumeAsBinaryAsync(consumerInstance, topic, maxBytes as int?, CancellationToken.None).ConfigureAwait(false);
}

public async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
    ConsumerInstance consumerInstance,
    string topic,
    int maxBytes,
    CancellationToken cancellationToken)
{
    return await ConsumeAsBinaryAsync(consumerInstance, topic, maxBytes as int?, CancellationToken.None).ConfigureAwait(false);
}

private async Task<ConfluentResponse<List<BinaryMessage>>> ConsumeAsBinaryAsync(
    ConsumerInstance consumerInstance,
    string topic,
    int? maxBytes,
    CancellationToken cancellationToken)
{
    string requestUri = BuildConsumeRequestUri(topic, maxBytes);

    HttpRequestMessage request = CreateRequestMessage(HttpMethod.Get, requestUri, consumerInstance.BaseUri)
        .WithContentType(ContentTypeKafkaBinary)
        .WithHostHeader(consumerInstance.BaseUri);

    return await ProcessRequest<List<BinaryMessage>>(request, cancellationToken).ConfigureAwait(false);
}
```

similarly for the other `int? maxBytes`

**Comment by bschmidtbauer** [comment]
Ok.  How would you want to handle scenarios like the test method `private int? GetMaxBytesOrNull()` then?  Maybe this is a scenario that isn't likely to occur in the wild but fixing the tests requires that it gets addressed.

The tests that are currently like the below where an actual value or null will work:

```
            Run(() => _confluentClient.ConsumeAsBinaryAsync(new ConsumerInstance
            {
                BaseUri = string.Format("{0}/consumers/{1}/instances/{2}", _baseUrl, textBoxConsumerGroup.Text, textBoxConsumerId.Text),
                InstanceId = textBoxConsumerId.Text
            }, textBoxTopic.Text, GetMaxBytesOrNull()).Result);
```

With the change to the interface that would no longer work and you would need to select the method to run at runtime - one path specifies the int value and the other calls the overload that doesn't require it.

```
            var maxBytes = GetMaxBytesOrNull();
            if (maxBytes.HasValue)
            {
                Run(() => _confluentClient.ConsumeAsBinaryAsync(new ConsumerInstance
                {
                    BaseUri = string.Format("{0}/consumers/{1}/instances/{2}", _baseUrl, textBoxConsumerGroup.Text, textBoxConsumerId.Text),
                    InstanceId = textBoxConsumerId.Text
                }, textBoxTopic.Text, maxBytes.Value).Result);
            }
            else
            {
                Run(() => _confluentClient.ConsumeAsBinaryAsync(new ConsumerInstance
                {
                    BaseUri = string.Format("{0}/consumers/{1}/instances/{2}", _baseUrl, textBoxConsumerGroup.Text, textBoxConsumerId.Text),
                    InstanceId = textBoxConsumerId.Text
                }, textBoxTopic.Text).Result);
            }
```

Thoughts?

**Comment by josephjeganathan** [comment]
Yes, that looks good in TestHarness. Please change it to `if (maxBytes.HasValue){ }else{ }` in the test harness as your last snippet shows.

**Comment by bschmidtbauer** [comment]
You should be all set - let me know if you need anything further.

**Comment by josephjeganathan** [comment]
:shipit: 

I'll add release note and ship a Nuget this weekend, thanks.

**Pull request merged**